boards/shared/Inc/Io/Io_SharedSpi.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include <stm32f3xx_hal.h>
+#include <stm32f3xx_hal_spi.h>
+
+struct SharedSpi;
+
+/**
+ * Allocate and initialize a SPI interface.
+ * @param spi_handle The handle of the given SPI interface.
+ * @param nss_port Pointer to the GPIO port of the given SPI interface.
+ * @param nss_pin The GPIO pin of the given SPI interface.
+ * @param delay_ms The delay in milliseconds during a SPI transaction which a
+ * blocking process should wait until an error is returned.
+ * @return Pointer to the allocated and initialized SPI interface
+ */
+struct SharedSpi *Io_SharedSpi_Create(
+    SPI_HandleTypeDef *spi_handle,
+    GPIO_TypeDef *     nss_port,
+    uint16_t           nss_pin,
+    uint32_t           delay_ms);
+
+/**
+ * Set the NSS pin low for the given SPI interface.
+ * @param spi The given SPI interface.
+ */
+void Io_SharedSpi_SetNssLow(const struct SharedSpi *spi);
+
+/**
+ * Set the NSS pin high for the given SPI interface.
+ * @param spi The given SPI interface.
+ */
+void Io_SharedSpi_SetNssHigh(const struct SharedSpi *spi);
+
+/**
+ * Transmit data to and receive data from the device connected to the given SPI
+ * interface.
+ * @param spi The given SPI interface.
+ * @param tx_data The data transmitted to the device connected to the SPI
+ * interface.
+ * @param tx_size The size of the data transmitted to the device connected to
+ * the SPI interface.
```

Comment:
So this is actually poorly named in the HAL_SPI library, it does not represent the size of the data but instead the amount of SPI messages/packets/frames to transmit/receive. Size implies bytes. Passing a _1U_  to this function would mean send/receive one SPI data frame which is of the size that the SPI peripheral is configured in (could be either 8bits or 16bits or less)

Reply:
Oh okay good point. I was just going by what was in the HAL library but sometimes that may not be the best idea lol

Reply:
Ya I found the HAL naming very confusing and so do a lot of people online, I had to go through the driver to figure out what this was actually doing and what is the correct value to pass to that parameter

Reply:
On a second though I guess it is referring to the size of the tx data or the rx buffer?

tx_data[tx_size]
rx_data[rx_size]

Reply:
hmmm I see should we re-word it to say that tx_size is the number of elements in the tx_data array ?

Reply:
Yup I can do that